numpy/_core/src/multiarray/nditer_pywrap.c

```diff
@@ -729,32 +717,60 @@ npyiter_init(NewNpyArrayIterObject *self, PyObject *args, PyObject *kwds)
         return -1;
     }
 
-    /* Set the dtypes and ops to all NULL to start */
-    memset(op_request_dtypes, 0, sizeof(op_request_dtypes));
+    /* Need nop to set up workspaces */
+    PyObject **op_objs = NULL;
+    PyObject *op_in_owned;  /* Sequence/object owning op_objs. */
+    int nop = npyiter_prepare_ops(op_in, &op_in_owned, &op_objs);
+    if (nop < 0) {
+        npy_free_cache_dim_obj(itershape);
```

Comment:
Maybe a `goto early_failure` (or `pre_alloc_fail`) that goes to the end of `finish`? Would require initializing `op_in_owned=NULL` (`res = -1` is already OK), so one could have,
```
...
    npy_free_workspace(op_axes);
early_fail:
    Py_XDECREF(op_in_owned);
    npy_free_cache_dim_obj(itershape);
    return res;
}
```
Could be used above and below.